File: Test cases/generics.swift
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// gryphon output: Test cases/Bootstrap Outputs/generics.swiftAST
// gryphon output: Test cases/Bootstrap Outputs/generics.gryphonASTRaw
// gryphon output: Test cases/Bootstrap Outputs/generics.gryphonAST
// gryphon output: Test cases/Bootstrap Outputs/generics.kt

struct Box<T> { // gryphon ignore
	let x: T
}

// gryphon insert: internal data class Box<T>(
// gryphon insert: 	val x: T
// gryphon insert: ) {
// gryphon insert: }

//// Instancing generic classes
let box = Box(x: 0)

//// Declaring generic functions
func f1<T>(box: Box<T>) {
	print(box.x)
}

// With generic return types
func f2<T>(box: Box<T>) -> T {
	return box.x
}

// In extensions
extension Box {
	func f3<U>(box: Box<U>) {
		print(self.x)
		print(box.x)
	}
}

//// Calling generic functions
f1(box: Box(x: 1))
print(f2(box: Box(x: 2)))
Box(x: 3).f3(box: Box(x: 4))

File: Test files/Test cases/generics.kt
//
// Copyright 2018 Vinicius Jorge Vendramini
//
// Licensed under the Hippocratic License, Version 2.1;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://firstdonoharm.dev/version/2/1/license
//
// To the full extent allowed by law, this software comes "AS IS,"
// WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED, and licensor and any other
// contributor shall not be liable to anyone for any damages or other
// liability arising from, out of, or in connection with the sotfware
// or this license, under any kind of legal claim.
// See the License for the specific language governing permissions and
// limitations under the License.
//
internal data class Box<T>(
	val x: T
)

internal open class Box2<T> {
	val x: T? = null
}

internal fun <T> f1(box: Box<T>) {
	println(box.x)
}

internal fun <T> f2(box: Box<T>): T {
	return box.x
}

internal fun <T> Box<T>.f3() {
}

internal fun <U, T> Box<T>.f4(box: Box<U>) {
	println(this.x)
	println(box.x)
}

internal val <T> Box<T>.a: Int
	get() {
		return 0
	}

internal fun <Key> f(): Map<Key, String> {
	return mapOf()
}

fun main(args: Array<String>) {
	val box: Box<Int> = Box(0)
	val box1: Box<Int> = Box(0)

	f1(Box(1))
	println(f2(Box(2)))
	Box(3).f4(Box(4))
}
